src/nps_gazebo_ros_gpu_sonar_single_beam.cpp

```diff
@@ -0,0 +1,271 @@
+    this->laser_scan_sub_.reset();
+}
+
+/* original code copies Gazebo _msg to ROS laser_msg and publishes 
+////////////////////////////////////////////////////////////////////////////////
+// Convert new Gazebo message to ROS message and publish it
+void NpsGazeboRosGpuSingleBeamSonar::OnScan(ConstLaserScanStampedPtr &_msg)
+{
+  // We got a new message from the Gazebo sensor.  Stuff a
+  // corresponding ROS message and publish it.
+  sensor_msgs::LaserScan laser_msg;
+  laser_msg.header.stamp = ros::Time(_msg->time().sec(), _msg->time().nsec());
+  laser_msg.header.frame_id = this->frame_name_;
+  laser_msg.angle_min = _msg->scan().angle_min();
+  laser_msg.angle_max = _msg->scan().angle_max();
+  laser_msg.angle_increment = _msg->scan().angle_step();
+  laser_msg.time_increment = 0;  // instantaneous simulator scan
+  laser_msg.scan_time = 0;  // not sure whether this is correct
+  laser_msg.range_min = _msg->scan().range_min();
+  laser_msg.range_max = _msg->scan().range_max();
+  laser_msg.ranges.resize(_msg->scan().ranges_size());
+  std::copy(_msg->scan().ranges().begin(),
+            _msg->scan().ranges().end(),
+            laser_msg.ranges.begin());
+  laser_msg.intensities.resize(_msg->scan().intensities_size());
+  std::copy(_msg->scan().intensities().begin(),
+            _msg->scan().intensities().end(),
+            laser_msg.intensities.begin());
+  this->pub_queue_->push(laser_msg, this->pub_);
+}
+*/
+
+/* NPS code copies much of Gazebo _msg to ROS laser_msg but only returns
+   one processed range and intensity value.
+*/
+////////////////////////////////////////////////////////////////////////////////
+// Convert new Gazebo message to ROS message and publish it
+void NpsGazeboRosGpuSingleBeamSonar::OnScan(ConstLaserScanStampedPtr &_msg)
+{
+  // We got a new message from the Gazebo sensor.  Stuff a
+  // corresponding ROS message and publish it.
+  sensor_msgs::LaserScan laser_msg;
+  laser_msg.header.stamp = ros::Time(_msg->time().sec(), _msg->time().nsec());
+  laser_msg.header.frame_id = this->frame_name_;
+  laser_msg.angle_min = _msg->scan().angle_min();
+  laser_msg.angle_max = _msg->scan().angle_max();
+  laser_msg.angle_increment = _msg->scan().angle_step();
+  laser_msg.time_increment = 0;  // instantaneous simulator scan
+  laser_msg.scan_time = 0;  // not sure whether this is correct
+  laser_msg.range_min = _msg->scan().range_min();
+  laser_msg.range_max = _msg->scan().range_max();
+
+  // ROS output message size is 1
+  laser_msg.ranges.resize(0);
+  laser_msg.intensities.resize(0);
+
+  // calculate range and intensity from Gazebo array
+  float angle = laser_msg.angle_min;
+  float intensity = 0.0;
+  float range = laser_msg.range_max - laser_msg.range_min;
+  auto range_it = _msg->scan().ranges().begin();
+  auto intensity_it = _msg->scan().intensities().begin();
+  while (range_it != _msg->scan().ranges().end())
+  {
+    // sum of f(intensity, angle)
+    intensity += *range_it * cos(10.0 * angle);
```

Comment:
I think it is worth a conversation on how to implement the beam pattern.  I'll write up something and we can iterate with Andi.

Reply:
Agreed.  This code is a template for Andi to start with.